assembly/src/ast/module.rs

```diff
@@ -1,321 +1,718 @@
+        &self.path
+    }
 
-    /// Clears the source locations from this module.
-    pub fn clear_locations(&mut self) {
-        self.local_procs.iter_mut().for_each(|p| p.clear_locations())
+    /// Get the namespace of this module, e.g. `std` in `std::math::u64`
+    pub fn namespace(&self) -> &LibraryNamespace {
+        self.path.namespace()
     }
 
-    // SERIALIZATION / DESERIALIZATION
-    // --------------------------------------------------------------------------------------------
+    /// Returns true if this module belongs to the provided namespace.
+    pub fn is_in_namespace(&self, namespace: &LibraryNamespace) -> bool {
+        self.path.namespace() == namespace
+    }
+
+    /// Get the module documentation for this module, if it was present in the source code the
+    /// module was parsed from
+    pub fn docs(&self) -> Option<Span<&str>> {
+        self.docs.as_ref().map(|spanned| spanned.as_deref())
+    }
 
-    /// Returns byte representation of this [ModuleAst].
+    /// Get the type of module this represents:
     ///
-    /// The serde options are NOT serialized - the caller must keep track of the serialization
-    /// options used.
-    pub fn write_into<R: ByteWriter>(&self, target: &mut R, options: AstSerdeOptions) {
-        // asserts below are OK because we enforce limits on the number of procedure and length of
-        // module docs in the module parser
-
-        // serialize docs
-        match &self.docs {
-            Some(docs) => {
-                assert!(docs.len() <= u16::MAX as usize, "docs too long");
-                target.write_u16(docs.len() as u16);
-                target.write_bytes(docs.as_bytes());
-            }
-            None => {
-                target.write_u16(0);
+    /// See [ModuleKind] for details on the different types of modules.
+    pub fn kind(&self) -> ModuleKind {
+        self.kind
+    }
+
+    /// Returns true if this module is an executable module.
+    #[inline(always)]
+    pub fn is_executable(&self) -> bool {
+        self.kind.is_executable()
+    }
+
+    /// Returns true if this module is a kernel module.
+    #[inline(always)]
+    pub fn is_kernel(&self) -> bool {
+        self.kind.is_kernel()
+    }
+
+    /// Returns true if this module has an entrypoint procedure defined,
+    /// i.e. a `begin`..`end` block.
+    pub fn has_entrypoint(&self) -> bool {
+        self.index_of(|p| p.is_main()).is_some()
+    }
+
+    /// Get an iterator over the procedures defined in this module.
+    ///
+    /// The entity returned is an [Export], which abstracts over locally-defined procedures and
+    /// re-exported procedures from imported modules.
+    pub fn procedures(&self) -> core::slice::Iter<'_, Export> {
+        self.procedures.iter()
+    }
+
+    /// Same as [Module::procedures], but returns mutable references.
+    pub fn procedures_mut(&mut self) -> core::slice::IterMut<'_, Export> {
+        self.procedures.iter_mut()
+    }
```

Comment:
Do we need to provide this method? Doesn't seem like it is used anywhere.

Reply:
We can probably get rid of it if we make the `procedures` field part of the public API, but I had anticipated making those fields an internal detail, and only exposing a specific set of operations in the form of methods like this.

There are a couple places (e.g. `VisitMut`) where we probably should be using this function rather than accessing the `procedures` field directly.

Reply:
This is now being used in at least one key place, and will probably get used more frequently over time, we just don't have too much rewrite stuff happening yet.